app/services/role_checker.rb

```diff
@@ -0,0 +1,48 @@
+class RoleChecker
+  attr_reader :user, :records
+
+  EDITOR_ROLES = %w(owner collaborator expert scientist moderator)
+  APPROVER_ROLES = %w(owner collaborator)
+  VIEWER_ROLES = %w(owner collaborator expert scientist tester)
+
+  def initialize(user, records)
+    @user = user
+    @records = records
+    @viewer_project_ids = viewer_project_ids
+  end
+
+  def can_edit?
+    ids = user_role_ids(user.roles, EDITOR_ROLES)
+    check_roles(ids, records)
+  end
+
+  def can_approve?
+    ids = user_role_ids(user.roles, APPROVER_ROLES)
+    check_roles(ids, records)
+  end
+
+  def can_view?
+    ids = user_role_ids(user.roles, VIEWER_ROLES)
+    check_roles(ids, records)
+  end
+
+  def viewer_project_ids
+    user_role_ids(user.roles, VIEWER_ROLES)
+  end
+
```

Comment:
methods below here should be private, they are implementation details that are used in the public api methods above. they should not be tested either.